include/zephyr/drivers/sensor_v2.h

```diff
@@ -0,0 +1,989 @@
+ * @return 0 on success
+ * @return < 0 on failure
+ */
+__syscall int sensor_set_bias(const struct device *sensor, uint32_t sensor_type, int8_t temperature,
+			      fp_t bias_x, fp_t bias_y, fp_t bias_z);
+
+static inline int z_impl_sensor_set_bias(const struct device *sensor, uint32_t sensor_type,
+					 int8_t temperature, fp_t bias_x, fp_t bias_y, fp_t bias_z)
+{
+	const struct sensor_driver_api_v2 *api = sensor->api;
+
+	if (api == NULL || api->set_bias == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->set_bias(sensor, sensor_type, temperature, bias_x, bias_y, bias_z);
+}
+
+/**
+ * @brief Update the sensor's resolution
+ *
+ * Set the sensor's resolution in terms of the number of bits per sample.
+ *
+ * @param sensor Pointer to the sensor device
+ * @param sensor_type The type of sensor to read
+ * @param resolution The new resolution measured in number of bits per sample
+ * @param round_up Whether or not to round up to the nearest available resolution
+ * @return 0 on success
+ * @return < 0 on failure
+ */
+__syscall int sensor_set_resolution(const struct device *sensor, uint32_t sensor_type,
+				    uint8_t resolution, bool round_up);
+
+static inline int z_impl_sensor_set_resolution(const struct device *sensor, uint32_t sensor_type,
+					       uint8_t resolution, bool round_up)
+{
+#ifdef CONFIG_SENSOR_RESOLUTION
+	const struct sensor_driver_api_v2 *api = sensor->api;
+
+	if (api == NULL || api->set_resolution == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->set_resolution(sensor, sensor_type, resolution, round_up);
+#else
+	return -ENOSYS;
+#endif
+}
+
+/**
+ * @brief Set the buffer to use for the next FIFO read
+ *
+ * @param sensor Pointer to the sensor device
+ * @param buffer The buffer to use
+ * @return 0 on success
+ * @return < 0 on failure
+ */
+__syscall int sensor_set_fifo_data_buffer(const struct device *sensor,
+					  struct sensor_raw_data *buffer);
+
+static inline int z_impl_sensor_set_fifo_data_buffer(const struct device *sensor,
+						     struct sensor_raw_data *buffer)
+{
+#ifdef CONFIG_SENSOR_STREAMING_MODE
+	const struct sensor_driver_api_v2 *api = sensor->api;
+
+	if (api == NULL || api->set_fifo_data_buffer == NULL) {
+		return -ENOSYS;
+	}
+
+	return api->set_fifo_data_buffer(sensor, buffer);
+#else
+	return -ENOSYS;
+#endif
+}
+/**
+ * @brief Set the processing function for new data.
+ *
+ * Set the sensor's function for processing data. This callback is important
+ * because sensors which make use of a FIFO should not also need to cache the data
+ * in memory while waiting. Instead, these sensors will rely on this callback to
+ * notify the application of the new data. If the application did not set the
+ * callback, the data will be discarded.
+ *
+ * @param sensor Pointer to the sensor device
+ * @param callback A callback function to use when new data is available
+ * @return 0 on success
+ * @return < 0 on failure
+ */
+__syscall int sensor_set_process_data_callback(const struct device *sensor,
```

Comment:
will this be substitute for old sensor trigger? if so this should not be considered only for streaming mode. It is also important to allow one shot sensor trigger.

Reply:
The thinking I had was that one shot sensors use the get API which is blocking. If you need a one off reading and require it to be non-blocking then you should manage your own thread (or use the sensor subsystem which will be coming quickly after this API). I'd like to avoid doing any threading in the sensor drivers (https://github.com/zephyrproject-rtos/zephyr/pull/44098#issuecomment-1171013737)

Reply:
Threading in sensors should be avoided but we cannot leave only a blocking API. The simple use case is using sensor as a wakeup source or action start.
If we don't want to handle callbacks directly from ISR that can be passed via system work queue.

At the moment we use sensor triggers in our applications. Example is here: https://github.com/nrfconnect/sdk-nrf/blob/main/applications/nrf_desktop/src/hw_interface/wheel.c#L76

Note that not all platforms can actually have resources to spawn another thread. For instance the mouse wheel above is suppose to work on nrf52810 which has very constraint memory. I expect people can be in even worse situation.
I highly recommend to have direct callbacks (triggers) from sensors (either from ISR or via system workq context).